Add ability to add streams and datasets in programs

gokulavasan opened the pull request:
JIRA : https://issues.cask.co/browse/CDAP-2777

Build : http://builds.cask.co/browse/CDAP-DUT2499-1

Note: This PR doesn't add any validation logic to make sure that datasets/streams with different configuration but with same names are used in two different programs. It will just use the last configuration for dataset/stream.

nitinmotgi commented on cdap-api/src/main/java/co/cask/cdap/api/app/ApplicationConfigurer.java:
```diff
@@ -35,72 +31,23 @@
-   * Adds a {@link Stream} to the Application.
-   *
-   * @param stream The {@link Stream} to include in the Application
-   */
```
We should not remove them. Please don't remove APIs this way. We have to deprecate and follow the standard process.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/DatasetConfigurer.java:
```diff
@@ -0,0 +1,77 @@
+   * @param typeName Name of the dataset type
+   * @param properties Dataset instance properties
+   */
+  @Beta
```
not related to your change, but it's weird that this is marked as beta

gokulavasan replied:
@sreevatsanraman @anew  Should we remove the ``@Beta`` annotation?

albertshau replied:
we can keep for now as its unrelated to this pr, but maybe open a jira to revisit beta apis. We tend to annotate as beta and never remove.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/flow/flowlet/AbstractFlowlet.java:
```diff
@@ -106,6 +110,70 @@ protected void setProperties(Map<String, String> properties) {
   /**
+   * @see FlowletConfigurer#addStream(Stream)
+   */
+  protected final void addStream(Stream stream) {
```
this code shows up in a bunch of different places (also true for existing setName, setDescription, setProperties methods)? Is there a reason we don't have them all extend some common base class?

albertshau replied:
actually, does it make sense to be able to add a stream in a flowlet? Connections between streams and flowlets has to happen at the flow level right?

gokulavasan replied:
@albertshau For the second question, the PR now allows Flows to add streams/datasets.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java:
```diff
@@ -90,6 +95,72 @@ protected final void setReducerResources(Resources resources) {
+   * @see MapReduceConfigurer#createDataset(String, Class, co.cask.cdap.api.dataset.DatasetProperties)
+   */
+  protected final void createDataset(String datasetName,
+                               Class<? extends Dataset> datasetClass,
```
nit: indent is off

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/AbstractMapReduce.java:
```diff
@@ -90,6 +95,72 @@ protected final void setReducerResources(Resources resources) {
+   * @see MapReduceConfigurer#createDataset(String, Class, DatasetProperties)
+   */
+  protected final void createDataset(String datasetName,
+                               Class<? extends Dataset> datasetClass) {
```
nit: indent is off

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceSpecification.java:
```diff
@@ -59,4 +62,19 @@
+  /**
+   * @return Map of dataset modules created in this program.
+   */
+  Map<String, String> getDataSetModules();
```
I think we've been replacing DataSet with Dataset

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceSpecification.java:
```diff
@@ -59,4 +62,19 @@
+  /**
+   * @return Map of dataset instances and {@link DatasetCreationSpec} created in this program.
+   */
+  Map<String, DatasetCreationSpec> getDataSetInstances();
```
these aren't really instances, maybe getDatasetSpecs() would be more accurate

albertshau replied:
are these methods useful? How would people use them?

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceSpecification.java:
```diff
@@ -59,4 +62,19 @@
+  /**
+   * @return Map of streams and {@link StreamSpecification} created in this program.
+   */
+  Map<String, StreamSpecification> getStreams();
```
these 3 methods also implemented in a bunch of places, is there a reason they aren't shared?

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/worker/AbstractWorker.java:
```diff
@@ -103,6 +107,72 @@ protected void useDatasets(Iterable<String> datasets) {
+   * @see WorkerConfigurer#createDataset(String, Class, co.cask.cdap.api.dataset.DatasetProperties)
+   */
+  protected final void createDataset(String datasetName,
+                               Class<? extends Dataset> datasetClass,
```
indentation

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/DefaultAppConfigurer.java:
```diff
@@ -149,6 +102,13 @@ public void addFlow(Flow flow) {
+
+    for (Map.Entry<String, FlowletDefinition> flowletEntry : spec.getFlowlets().entrySet()) {
+      FlowletSpecification flowletSpec = flowletEntry.getValue().getFlowletSpec();
+      streams.putAll(flowletSpec.getStreams());
```
add a todo for checking for conflicts

albertshau commented:
mostly questions about repeat code

gokulavasan commented:
@albertshau Please review.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/flow/flowlet/FlowletSpecification.java:
```diff
@@ -178,7 +196,10 @@ public AfterDescription withResources(Resources resources) {
         return new DefaultFlowletSpecification(name, description, failurePolicy,
-                                               dataSets.build(), arguments, resources);
+                                               dataSets.build(), arguments, resources,
+                                               Maps.<String, StreamSpecification>newHashMap(),
```
we want to eventually remove guava usage in api, better not to add new usages.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/internal/api/ProgramDataset.java:
```diff
@@ -0,0 +1,97 @@
+import co.cask.cdap.api.dataset.module.DatasetModule;
+
+/**
+ *
```
add a javadoc

albertshau commented on cdap-api/src/main/java/co/cask/cdap/internal/api/ProgramDataset.java:
```diff
@@ -0,0 +1,97 @@
+/**
+ *
+ */
+public class ProgramDataset {
```
this is an odd name, makes it sound like a dataset instead of a program. DatasetProgram? DataProgram? AbstractProgram?

gokulavasan replied:
Oh I wanted to change the name of this class. Deferred the naming. Will choose a better name.

chtyim replied:
Need to be abstract.

chtyim replied:
This class should name something like `AbstractProgramDatasetConfigurable`.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/worker/WorkerSpecification.java:
```diff
@@ -19,7 +19,9 @@
 import co.cask.cdap.api.common.PropertyProvider;
+import co.cask.cdap.api.data.stream.StreamSpecification;
 import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.internal.dataset.DatasetCreationSpec;
```
this should be in api if we're exposing it here.

albertshau commented on cdap-api/src/main/java/co/cask/cdap/internal/api/ProgramDataset.java:
```diff
@@ -0,0 +1,97 @@
+ * the License.
+ */
+
+package co.cask.cdap.internal.api;
```
how come this is internal?

gokulavasan replied:
@albertshau Hmm, should I move this class to co.cask.cdap.api package?

gokulavasan replied:
@chtyim Any suggestion on which package this class needs to go into?

albertshau replied:
Was originally thinking it should be since a bunch of things in api extend it.  But thinking about it a little more, we don't want anybody extending it so maybe it should be in internal.

albertshau commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/ApplicationSpecification.java:
```diff
@@ -75,15 +78,13 @@
   /**
-   * @return An immutable {@link Map} from {@link co.cask.cdap.api.dataset.module.DatasetModule} name
-   *         to {@link co.cask.cdap.api.dataset.module.DatasetModule} class name for
+   * @return An immutable {@link Map} from {@link DatasetModule} name to {@link DatasetModule} class name for
```
An unmodifiable

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/DatasetConfigurer.java:
```diff
@@ -0,0 +1,77 @@
+   *
+   * @param stream {@link Stream}
+   */
+  void addStream(Stream stream);
```
Should have a `addStream(String streamName)` as well.

gokulavasan replied:
@chtyim Why do we need that? This ``addStream(Stream stream)`` will create a stream if it doesn't exist. What will ``addStream(String streamName)`` do?

chtyim replied:
It's just a overloaded method that calls `addStream(new Stream(streamName))`.

gokulavasan replied:
Oh ok. Makes sense. Will add that method. Thanks @chtyim

gokulavasan replied:
@chtyim Do we need this in both the Configurer and the ``Abstract{Flow, Spark, MapReduce etc}`` ? Or will just the latter suffice?

chtyim replied:
That's ok.

albertshau commented on cdap-proto/src/main/java/co/cask/cdap/proto/codec/MapReduceSpecificationCodec.java:
```diff
@@ -77,8 +83,22 @@ public MapReduceSpecification deserialize(JsonElement json, Type typeOfT, JsonDe
     Map<String, String> properties = deserializeMap(jsonObj.get("properties"), context, String.class);
 
+    JsonElement streamElement = jsonObj.get("streams");
+    Map<String, StreamSpecification> streams = (streamElement == null) ?
```
does deserializeMap not handle the null case?

gokulavasan replied:
Hmm, looks like it does. I will remove the unnecessary checks.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/DatasetConfigurer.java:
```diff
@@ -0,0 +1,77 @@
+   * @param properties Dataset instance properties
+   */
+  @Beta
+  void createDataset(String datasetName, String typeName, DatasetProperties properties);
```
Shall we also have variant that doesn't take `DatasetProperties` so that caller doesn't need to pass in `DatasetProperties.EMPTY`?

gokulavasan replied:
Sure, I will add that.

gokulavasan replied:
@chtyim Just realized that the helper methods in Abstract{Flow, Spark, Flowlet, MapReduce} etc provide a method that doesn't need the caller to pass in ``DatasetProperties.EMPTY``. So should we add a method in the Configurer as well?

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/ProgramConfigurer.java:
```diff
@@ -0,0 +1,47 @@
+  void setDescription(String description);
+
+  /**
+   * Sets a set of properties that will be available through the {@link ProgramSpecification's getProperties()}
```
This is not very accurate since `ProgramSpecification` doesn't have a `getProperties()` method.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/flow/FlowSpecification.java:
```diff
@@ -74,6 +76,21 @@
   /**
+   * @return Map of streams and {@link StreamSpecification} created in this flow.
+   */
+  Map<String, StreamSpecification> getStreams();
```
Why need these methods?

gokulavasan replied:
We need to use these methods for use in the ``DefaultAppConfigurer`` to get the streams, datasetModules, datasetSpecs from each program.

chtyim replied:
The configurer is the one who create the spec, right? Why it is reversed?

gokulavasan replied:
True. I will move these methods into the configurers and remove them from the Program specifications since there is no use case to expose this to the users. That will also take care of the repeated code in specifications.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/flow/flowlet/FlowletSpecification.java:
```diff
@@ -76,6 +79,21 @@
   /**
+   * @return Map of streams and {@link StreamSpecification} created in this flowlet
+   */
+  Map<String, StreamSpecification> getStreams();
```
Why get repeat again? If these methods are needed in the specification, shall we have a super-interface for these methods?

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/mapreduce/MapReduceSpecification.java:
```diff
@@ -124,6 +134,27 @@ public Resources getReducerResources() {
+  /**
+   * @return Map of streams and {@link StreamSpecification} created in this program.
+   */
+  public Map<String, StreamSpecification> getStreams() {
```
should we put these in ProgramSpecification?

chtyim replied:
Repeated the 3rd time.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/internal/api/ProgramDataset.java:
```diff
@@ -0,0 +1,97 @@
+ */
+public class ProgramDataset {
+
+  protected DatasetConfigurer datasetConfigurer;
```
should be private final and initialize it through constructor. It's generally bad practice to have child class to setup this directly.

gokulavasan replied:
True. But the datasetConfigurer will not be available when this is class' constructor is invoked. I will see if we can do a better job at structuring this.

chtyim replied:
Making this class abstract and have an abstract method `<T extends DatasetConfigurer> T getConfigurer()`.

gokulavasan replied:
Thanks @chtyim. Will make that change.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/internal/api/ProgramDataset.java:
```diff
@@ -0,0 +1,97 @@
+   *
+   * @see DatasetConfigurer#createDataset(String, Class, co.cask.cdap.api.dataset.DatasetProperties)
+   */
+  protected final void createDataset(String datasetName, Class<? extends Dataset> datasetClass,
```
Why this and next methods are not `@Beta`? I get confused what's the rule of having `@Beta`.

gokulavasan replied:
@chtyim I just followed the notation used in ``AbstractApplication``

chtyim replied:
Meaning the existing one is also not good?

gokulavasan replied:
@chtyim That is possible. I was not sure about the Beta notations myself. I will open a JIRA to revisit these and remove them as necessary.

gokulavasan commented:
@chtyim @albertshau Thanks a lot for the review! The PR is much smaller now. Please take a look again. Specifically these three comments : 
https://github.com/caskdata/cdap/pull/3179#discussion_r36922144
https://github.com/caskdata/cdap/pull/3179#discussion_r36922245
https://github.com/caskdata/cdap/pull/3179#discussion_r36923279

chtyim commented on cdap-api/src/main/java/co/cask/cdap/internal/api/AbstractProgramDatasetConfigurable.java:
```diff
@@ -0,0 +1,106 @@
+ *
+ * @param <T> Program's configurer
+ */
+public abstract class AbstractProgramDatasetConfigurable<T extends DatasetConfigurer> {
```
Should be ok inside internal package. For classes that are in internal packages, they are needed by classes in cdap-api modules, but not part of public API. Meaning we don't need to maintain compatibility for those classes.

gokulavasan replied:
@chtyim But Abstract{Flow, MapReduce, etc} extend this class. So will that violate the compatibility guarantees?

chtyim replied:
No. Why?

gokulavasan replied:
Say if we modify some method signature in ``AbstractProgramDatasetConfigurable`` that will affect ``Abstract{Flow, MR, Spark}`` etc. So the same constraints of deprecating API will apply to this class as well even though it is in co.cask.cdap.internal.api.

albertshau replied:
is this where the plugin methods will eventually go?  Maybe we should just call it AbstractConfigurableProgram?  Don't have a strong opinion on this though

chtyim replied:
Leaving it in internal so that we can rename the class or even the hierarchy later (as long as the methods are still compatible).

chtyim commented on cdap-api/src/main/java/co/cask/cdap/internal/api/DefaultDatasetConfigurer.java:
```diff
@@ -0,0 +1,119 @@
+/**
+ * Implementation of {@link DatasetConfigurer} for adding datasets and streams in Configurers.
+ */
+public class DefaultDatasetConfigurer implements DatasetConfigurer {
```
Implementation of configurer shouldn't be in cdap-api. It should be inside cdap-app-fabric.

gokulavasan replied:
@chtyim True, but the ``DefaultFlowletConfigurer`` is in cdap-api which depends on ``DefaultDatasetConfigurer``. ``DefaultFlowConfigurer`` is in cdap-api since ``FlowletDefinition`` needs it. And ``FlowletDefinition`` is in cdap-api since the old configurer style ``FlowSpecification`` needs it. So for now I have placed it in cdap.internal.api. Will move it to cdap-app-fabric after removing old Flow configurer API.

chtyim replied:
The idea of Configurer interface is that only the interface is in cdap-api so that we are free to implement it inside cdap-app-fabric without worrying about bring in extra library dependencies (which we cannot do with the old configure() method API).

gokulavasan replied:
Thanks @chtyim! So for this comment should I just make a note of it to move the class to cdap-app-fabric in the 'Remove old Flow API' ?

chtyim replied:
Add a TODO for this class with the JIRA mentioned.

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java:
```diff
@@ -191,6 +203,27 @@ public int getInstances() {
   /**
+   * @return Map of Stream name and {@link StreamSpecification} created in this Flowlet.
+   */
+  public Map<String, StreamSpecification> getStreams() {
```
Are these methods only needed for old Flow configuration API compatibility? Because when we switched to configurer way, I think these methods shouldn't be needed.

gokulavasan replied:
These are required because in ``DefaultFlowConfigurer`` we use ``FlowletDefinition`` to create and get ``FlowletSpecification`` instead of directly using ``FlowletConfigurer``. Hence we had to add these methods to access the streams/datasets created inside that Flowlet.

chtyim replied:
That seems wrong and defeated the purpose of using configurer approach.

gokulavasan commented:
@chtyim Added a note to move the classes and remove the methods once the old Flow API is removed. Please let me know if there is anything else that be fixed. Thanks for the review!

albertshau commented on cdap-api/src/main/java/co/cask/cdap/api/flow/FlowSpecification.java:
```diff
@@ -18,9 +18,11 @@
 
 import co.cask.cdap.api.ProgramSpecification;
 import co.cask.cdap.api.data.stream.Stream;
+import co.cask.cdap.api.data.stream.StreamSpecification;
```
unneeded?

albertshau commented:
lgtm, @chtyim did your comments get addressed?

chtyim commented on cdap-api/src/main/java/co/cask/cdap/api/flow/FlowletDefinition.java:
```diff
@@ -190,6 +202,29 @@ public int getInstances() {
+  /**
+   * @return Map of Stream name and {@link StreamSpecification} created in this Flowlet.
+   */
+  public Map<String, StreamSpecification> getStreams() {
```
Probably mark these methods as `@Deprecate` to make sure user don't call them.

chtyim commented:
Just a few comments about comment and deprecation left. Rest LGTM. Also, better to squash into one commit before merging since there are so many tiny commits for this PR.

gokulavasan commented:
@chtyim Addressed your comment about deprecation. Do I have anything unaddressed? I will squash all my commits.

gokulavasan commented:
@albertshau @chtyim Addressed all comments. Created this new PR that contains the diff from this PR : https://github.com/caskdata/cdap/pull/3593
It was difficult to squash all commits after merges in between. I will close this PR and merge the other one once I get a LGTM on the new one.
